### PR TITLE
HubConnection implements IAsyncDisposable

### PR DESCRIPTION
**Pull request by BrennanConroy**
Fixes https://github.com/aspnet/AspNetCore/issues/13082

### CODE REVIEW

File: src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs [hunk 1]
```diff
@@ -504,8 +503,15 @@ private async Task StopAsyncCore(bool disposing)
 
                 if (disposing)
                 {
-                    (_serviceProvider as IDisposable)?.Dispose();
                     _disposed = true;
+                    if (_serviceProvider is IAsyncDisposable asyncDispose)
```
Comment (BrennanConroy) [inline src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs]: This is fun, `_serviceProvider` calls `HubConnection.DisposeAsync` so we need to set the bool first.

And for netstandard2.0 we need to keep `IDisposable` I believe.

File: src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs [hunk 2]
```diff
@@ -1528,6 +1534,11 @@ private static void SafeAssert(bool condition, string message, [CallerMemberName
             }
         }
 
+        ValueTask IAsyncDisposable.DisposeAsync()
```
Comment (BrennanConroy) [inline src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs]: We need to keep the Task returning DisposeAsync to avoid making a breaking change

**Comment by anurse** [comment]
Let's retarget this to `master` to get it in and open a new issue for backporting. I'm not 100% confident I can get everything in place before going on leave and would like it in somewhere.

File: src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs (hunk 2 above)
Comment (halter73) [reply src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs]: 😭

File: src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs (hunk 1 above)
Comment (halter73) [reply src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs]: Add a comment about setting the bool in the correct order.

**Review by halter73 — APPROVED** [review]

**Pull request closed without merging**